ocaml/xapi/repository.ml

```diff
@@ -0,0 +1,127 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module Unixext = Xapi_stdext_unix.Unixext
+
+module D = Debug.Make (struct let name = "repository" end)
+
+open D
+
+let cmd_yum = "/usr/bin/yum"
+let cmd_rm = "/usr/bin/rm"
+
+let reposync_mutex = Mutex.create ()
+
+let create_repository_record ~__context ~name_label ~name_description ~binary_url ~source_url =
+  let ref = Ref.make () in
+  let uuid = Uuidm.to_string (Uuidm.create `V4) in
+  Db.Repository.create
+    ~__context
+    ~ref
+    ~uuid
+    ~name_label
+    ~name_description
+    ~binary_url
+    ~source_url
+    ~hash:""
+    ~up_to_date:false;
+  ref
+
+let assert_name_is_valid ~name_label =
+  match Re.Str.string_match (Re.Str.regexp "^[A-Za-z]+[A-Za-z0-9\\-]+$") name_label 0 with
+  | true -> ()
+  | false -> raise Api_errors.(Server_error (invalid_repository_name, [name_label]))
+
+let assert_url_is_valid ~url =
+  match !Xapi_globs.repository_domain_name_allowlist with
+  | [] -> ()
+  | l ->
+    let result = l |> List.exists (fun domain_name ->
+        let r = Re.Str.regexp
+            (Printf.sprintf
+               "^https?://[A-Za-z0-9\\-]+\\.%s\\(:[0-9]+\\)?/[A-Za-z0-9\\-/\\$]+$"
+               domain_name)
+        in
+        Re.Str.string_match r url 0)
+    in
+    begin match result with
+      | true -> ()
+      | false -> raise Api_errors.(Server_error (invalid_base_url, [url]))
+    end
+
+let introduce ~__context ~name_label ~name_description ~binary_url ~source_url =
+  assert_name_is_valid ~name_label;
+  assert_url_is_valid ~url:binary_url;
+  assert_url_is_valid ~url:source_url;
+  match Db.Repository.get_by_name_label ~__context ~label:name_label with
+  | [] ->
+    create_repository_record  ~__context ~name_label ~name_description ~binary_url ~source_url
+  | _ ->
+    error "A repository with same name_label '%s' already exists" name_label;
+    raise Api_errors.(Server_error (repository_already_exists, []))
```

Comment:
You could add the ref as an argument (also add it to the definition in the datamodel).